### ISSUE TITLE
Provide way of incrementing patch number with date-based versioning

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2915]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2915|
|Type|enhancement|
|Created|2007-11-20 18:44:54Z|
|Branch|https://github.com/twisted/twisted/tree/auto-version-2915-3|

Currently we do date-based versioning (i.e. YEAR.INCREMENT), but our version objects have fields for point-releases (X.Y.Z). Right now, 'Z' is always 0.

One day we might want it to be non-zero.

<details><summary>Searchable metadata</summary>

```
trac-id__2915 2915
type__enhancement enhancement
reporter__jml jml
priority__low low
milestone__totally_automated_release_infrastructure totally automated release infrastructure
branch__branches_auto_version_2915_3 branches/auto-version-2915-3
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1195584294000000 1195584294000000
changetime__1360402838000000 1360402838000000
version__None None
owner__therve therve
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

We very recently did 8.0.1.  I suppose that means we managed to make it non-zero.  Does that mean this ticket is resolved?  I don't think I quite understand the description.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

It's definitely a poor description.

I think it's meant to refer to `t.python._releases.getNextVersion` (pasted below), which quite clearly doesn't support patch numbers.

How did we increment the patch number in the 8.0.1 release?

```python
def getNextVersion(version, now=None):
    """
    Calculate the version number for a new release of Twisted based on
    the previous version number.

    @param version: The previous version number.
    @param now: (optional) The current date.
    """
    # XXX: This has no way of incrementing the patch number. Currently, we
    # don't need it. See bug 2915. Jonathan Lange, 2007-11-20.
    if now is None:
        now = date.today()
    major = now.year - VERSION_OFFSET
    if major != version.major:
        minor = 0
    else:
        minor = version.minor + 1
    return Version(version.package, major, minor, 0)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Thanks, that clears things up.  Maybe radix can comment on how 8.0.1 was produced.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix commented|
|-|-|

We haven't even used getNextVersion yet. I just call the function that takes a version number directly.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Are we planning to call `getNextVersion` at some point in the future?

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix commented|
|-|-|

Maybe once it provides a useful result :-)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Okay.  So presumably `getNextVersion` needs to have two different behaviors.  One behavior is that it increments the `major.minor` portion of a version in an appropriate manner, setting `patch` to 0.  The other is that it leaves `major.minor` alone and increments `patch`.  The former behavior is for new releases (ie, the first release of a new branch of trunk).  The latter is for subsequent releases of such branches.  Can it tell what is going on by itself?  Does it need a new parameter?  Should there just be two functions?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

I'd go for two different methods, possibly renaming the pair to `incrementRelease` and `incrementPatch` or something like that.

My working assumption is that the hypothetical automated release tool probably needs to take an argument as to whether a release is a bugfix release or not.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [36614]) Branching to 'auto-version-2915'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [36727]) Branching to 'auto-version-2915-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner|
|-|-|

The branch attached changes the way change-versions works: instead of passing the version, it takes 2 flags, prerelease and patch, which choose automatically the new version. Most of the changes are in getNextVersion, then ChangeVersionsScript and changeAllProjectVersions. I also removed updateTwistedVersionInformation which was not used anywhere.

getNextVersion code is fairly ugly, but the tests cover it properly, and match the use cases I have when doing releases.

Sorry for the formatting changes, I got a little bit carried away.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

I [forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/auto-version-2915-2) to make the reviewer's job easier.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_16"></a>|@Julian commented|
|-|-|

I also think this would be less ugly moved into two functions rather than writing `getNextVersion(version, False, False, now)`.

Is there a reason you preferred one?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

Well there are actually 3 versions, considering the prerelease bit, which overlaps with each other, so I don't think having 2 functions would really improve things.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince set owner to @therve|
|-|-|

  1. From twistedchecker: `ChangeVersionsScriptOptions` has no docstring.
  1. `getNextVersion` has missing descriptions of a couple of arguments.
  1. The code would be clearer if keyword arguments were used for `patch` and `prerelease`.

Otherwise looks good. Please address the above and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve commented|
|-|-|

(In [37081]) Branching to 'auto-version-2915-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set status to `closed`|
|-|-|

(In [37084]) Merge auto-version-2915-3

Author: therve
Reviewer: tom.prince
Fixes: [#2915](https://github.com/twisted/twisted/issues/2915)

Automatically pick the next version in the change-versions release script.
